lib/AST/Decl.cpp

```diff
@@ -3613,7 +3632,8 @@ bool EnumDecl::isFormallyExhaustive(const DeclContext *useDC) const {
   // Testably imported enums are exhaustive, on the grounds that only the author
   // of the original library can import it testably.
   if (auto *useSF = dyn_cast<SourceFile>(useDC->getModuleScopeContext()))
-    if (useSF->hasTestableImport(containingModule))
+    if (useSF->hasTestableOrPrivateImport(AccessLevel::Internal,
+                                          containingModule))
```

Comment:
This works for either private or testable imports. the access level is used inside of the function to determine whether we need to check the filename (if it is private or file private).

Using AccessLevel::Internal means we will use the check that checks whether an private or testable import occurred (without checking the filename in the private case).

Reply:
Uh, hm. But this should be specific to the enum we're asking the question of, right?

Reply:
You are correct it would make a difference for private imports and private/fileprivate declarations.

What you don't see in the diff is that there is a test before this test that checks if the enum is not public.

```
// Non-public, non-versioned enums are always exhaustive.
  AccessScope accessScope = getFormalAccessScope(/*useDC*/nullptr,
                                                 /*respectVersioned*/true);
  if (!accessScope.isPublic())
    return true;
```

We end up at ``useSF->hasTestableOrPrivateImport`` only with a public enum.

In this case we end up in the internal/public branch that does not depend on the file.

```
bool SourceFile::hasTestableOrPrivateImport(
    AccessLevel accessLevel, const swift::ValueDecl *ofDecl) const {
  auto *module = ofDecl->getModuleContext();
  switch (accessLevel) {
  case AccessLevel::Internal:
  case AccessLevel::Public:
    // internal/public access only needs an import marked as @_private. The
    // filename does not need to match (and we don't serialize it for such
    // decls).             
    return std::any_of(    
        Imports.begin(), Imports.end(),
        [module](ImportedModuleDesc desc) -> bool {
          return desc.module.second == module &&
                 (desc.importOptions.contains(ImportFlags::PrivateImport) ||
                  desc.importOptions.contains(ImportFlags::Testable));
        });
  case AccessLevel::Open:         
    return true;
  case AccessLevel::FilePrivate:
  case AccessLevel::Private:
    // Fallthrough.
    break;
  }    
```

Therefore this check is sufficient.

Reply:
Oof, right. Okay. Too many components depending on each other…